Fixed #34429 -- Allowed setting unusable passwords for users in the auth forms.

fsbraun opened the pull request:
https://code.djangoproject.com/ticket/34429

## Feature added
This PR introduces a new functionality to the admin change password form: the checkbox (on by default) "Allow login"

<img width="715" alt="image" src="https://github.com/django/django/assets/16904477/96cfe15f-8a92-4d73-9d8c-44c9e0c0d584">

If unchecked, the password field disappears and the form sets an unusable password upon submission. The user is warned before.

<img width="715" alt="image" src="https://github.com/django/django/assets/16904477/9c651847-2101-4167-9161-fb8022cee851">

## Tests

Existing tests have been changed to contain the new form field.

A new test was added to test the functionality.

## Other effects
This PR adds three new strings which will have to be translated.

ngnpope commented on django/contrib/admin/templates/admin/auth/user/change_password.html:
```diff
@@ -54,4 +70,5 @@
 
 </div>
 </form></div>
+<script src="{% static 'admin/js/change_password_form.js' %}" async></script>
```
Stick this in the `extrahead` block:

```html
{% block extrahead %}
    {{ block.super }}
    <script src="{% static 'admin/js/change_password_form.js' %}" async></script>
{% endblock %}
```

ngnpope commented on django/contrib/admin/templates/admin/auth/user/change_password.html:
```diff
@@ -28,8 +28,24 @@
 
+<ul id="id_unusable_warning" class="messagelist">
+  <li class="warning">
+    {% translate "Upon submission this will set a unusable password and prevent the user from logging in." %}
```
```suggestion
    {% translate "Upon submission an unusable password will be set which prevents the user from logging in." %}
```

ngnpope reviewed (changes requested):
Thanks for this, I've made some comments.

Please can you update the PR title to match the format in the contributing guidelines:

```
Fixed #34429 -- Allowed setting unusable password via admin UI.
```

This will also need documentation changes and release notes adding.

When you've done this you can uncheck `Needs documentation` and `Patch needs improvement` on ticket-34429.

ngnpope commented on tests/auth_tests/test_forms.py:
```diff
@@ -1371,3 +1402,21 @@ def test_html_autocomplete_attributes(self):
+            valid = form.is_valid()
+            form.save(commit=True)
+
+        self.assertTrue(valid)  # Valid despite password empty/mismatch
```
```suggestion
        self.assertIs(valid, True)  # Valid despite password empty/mismatch.
```

ngnpope commented on tests/auth_tests/test_forms.py:
```diff
@@ -1371,3 +1402,21 @@ def test_html_autocomplete_attributes(self):
+            form.save(commit=True)
+
+        self.assertTrue(valid)  # Valid despite password empty/mismatch
+        set_unusable_password.assert_called_once()
```
Instead of mocking, can we just use the following:

```suggestion
        self.assertIs(user.has_usable_password(), False)
```

fsbraun commented:
@ngnpope Thank you for the suggestions! Let me briefly comment and ask one more question:

* Probably we do not need to wait for `DOMContentLoaded` since the script is loaded as `async`. It would probably miss the event (which might be fired before it is loaded). I'll remove the event in total.
* Using the `hidden` property might clash with custom css which might set something like `.form-row { display: block }` effectively disabling the `hidden` property. OK, if I stay with `style.display`?

Will, of course, also make the other changes.

fsbraun commented:
One more thing to add: I did use `function () { ... }` instead of `() => { ... }` since otherwise `this` is not defined properly.

xi commented on django/contrib/auth/forms.py:
```diff
@@ -469,42 +469,67 @@ class AdminPasswordChangeForm(forms.Form):
         help_text=_("Enter the same password as before, for verification."),
     )
+    usable_password = forms.BooleanField(
+        label=_("Allow login?"),
```
I am not sure I like this label. Setting an unusable login is also an option if you use a different authentication backends.

fsbraun replied:
@xi What label would you like? Would a different authentication backend allow logins with an unusable login? If so, this would need explanation.

Would "Allow login with password?" be better? Help text then could be "If unchecked, the user will not be able to log in using their password."

xi replied:
I like your suggestions :+1:

shangxiao commented:
I have a remark about the wording 

> Allow login?
> If unchecked, the user will not be able to log in.

I think people are going to be confused about what the difference is between this and the active flag on the edit screen:

> Active.
> Designates whether this user should be treated as active. Unselect this instead of deleting accounts.

Nick I know you requested the addition of ? and I'm not saying it's good or bad but it is now inconsistent with other checkboxes (I'd just leave it off personally) 😛

shangxiao commented:
Can anything be done to stop the jumping checkbox? 🤔

https://github.com/django/django/assets/1845938/fe192289-704e-4db0-9f6f-9667d913c698

shangxiao commented:
Anyhoo nice work so far fsbraun!

fsbraun commented:
@shangxiao Thanks very much for the comments:
* The jumping can be stopped by putting the warning below the checkbox and above the submit button
* Thanks for pointing out the "active" alternative. Really, an unusable password only prevents login using username and password (as @xi probably also pointed out). I tried to better name those.

![login](https://github.com/django/django/assets/16904477/aa87271b-fd99-44f4-b012-4d46ab70f854)

fsbraun commented on django/contrib/admin/templates/admin/auth/user/change_password.html:
```diff
@@ -30,6 +34,25 @@
+  <li class="warning">
+    {% blocktranslate %}
+      Upon submission an unusable password will be set which prevents the user from logging in
+      using his username and password. Potential other means of authentication remain unaffected.
```
Right! I try to pay attention to those things but they still slip through 🤦‍♂️

fsbraun commented:
Thanks, everybody, for the constructive feedback. I have updated the PR:

* @knyghty The help text now reads
  
   _Upon submission an unusable password will be set which prevents the user from logging in
   using their username and password. The user's current password will be lost. Potential other
   means of authentication remain unaffected._ 
 
* @xi This is conveniently gender-neural and clarifies that the current password will be lost

* @knyghty I have moved the warning into the help_text div. This implies that once the form field sets the `aria-describedby`  property, it will include the help text **and** the warning. No need for two id to be provided to `aria-describedby` (which would be possible as far as I understand the docs).

fsbraun commented:
Is there anything I can do to move this forward?

felixxm commented on django/contrib/admin/templates/admin/auth/user/change_password.html:
```diff
@@ -30,6 +34,28 @@
+    <label class="vCheckboxLabel" for="id_usable_password">{{ form.usable_password.label }}</label>
+  </div>
+  {% if form.usable_password.help_text %}
+  <div class="help"{% if form.usable_password.id_for_label %} id="{{ form.usable_password.id_for_label }}_helptext">{% endif %}
```
">" should be outside of `{% if ... %}...{% endif %}`/

fsbraun replied:
Indeed! 🤦‍♂️

felixxm commented on tests/auth_tests/test_forms.py:
```diff
@@ -1371,3 +1402,20 @@ def test_html_autocomplete_attributes(self):
+        form.save(commit=True)
+
+        self.assertIs(valid, True)  # Valid despite password empty/mismatch.
+        self.assertIs(user.has_usable_password(), False)
```
Can we also add assertions for `password`.

fsbraun replied:
The only assertion I can think of is that `user.password` starts with `UNUSABLE_PASSWORD_PREFIX`. I feel this is somewhat redundant to `user.has_usable_password()` being `False´, but certainly, I can add this. Or were you thinking of something else?

felixxm replied:
Right, probably not needed.

felixxm reviewed (commented):
@fsbraun Thanks :+1: I left initial comments.

According to the ticket description we should have the same feature when adding a new user:
![image](https://github.com/django/django/assets/2865885/68bd9319-1916-4425-9763-ce7e89313246)

fsbraun commented on django/contrib/auth/forms.py:
```diff
@@ -469,42 +469,70 @@ class AdminPasswordChangeForm(forms.Form):
         ),
         strip=False,
+        required=False,  # Can be empty if usable_password is False
+        validators=[],  # Password validation happens in the check_password1() method
```
Yes, it's both needed and related. The password validation can only happen in the form's clean method, since a password is only required or needs to be validated if `has_usable_password` is `True`.

fsbraun commented:
@felixxm Thank you for your detailed feedback. Indeed, it overlooked that this should also apply to the user add form. This is fixed now:
* Since the change password and user add form are quite different (form with template and an admin form) I have refactored the HTML to avoid repetition and make it usable for both forms.
* Indentation has been fixed
* The initial value of `usable_password` now is `has_usable_password()`
* I annotated the docs with `versionchanged`
* Updated the docstrings for the tests
* Added a Selenium test
* Fixed the HTML layout issue (wrongly placed `>`)
Please let me know if there is anything else I can do!

fsbraun commented:
@felixxm Is there anything I can do to get this PR over the finish line? Would you like to take a final look?

nessita commented:
> I think people are going to be confused about what the difference is between this and the active flag on the edit screen:

I started reviewing this PR and the first think I thought was exactly this. Was there a consideration of being more "honest" with the admin user operating the form, something like:

Label: "Unset the password"
Help text: "Unsetting the password will render the login credentials unusable. The User will no longer be able to log in until a new password is set."

fsbraun commented:
@nessita Thanks for taking the time to look through this. 

Indeed, the wording has been discussed [here](https://github.com/django/django/pull/16942#discussion_r1218526204) (and I also tried to get some input [here](https://code.djangoproject.com/ticket/34429#comment:6)).

I did consider "Set unusable password" but this would require the user to understand the implications, namely, that the user cannot log in using a username and password. That's the root of the wording.

IMHO, "unsetting" a password could be confused with "set no password". Also, even with an unusable password, you can potentially still log in with a different authentication backend.

Current text: 
**Allow login with username and password**

Help text: 
**If unchecked, the user will have an unusable password and not be able to log in using their username and any password.** (I just realized this had a spelling error)

Warning: 
**Upon submission, an unusable password will be set which prevents the user from logging in
           using their username and password. The user's current password will be lost. Potential other
           means of authentication remain unaffected.**

I am not hung up on this, but also would like it to be as clear as possible. What's your call?

nessita commented:
> @nessita Thanks for taking the time to look through this.
> 
> Indeed, the wording has been discussed [here](https://github.com/django/django/pull/16942#discussion_r1218526204) (and I also tried to get some input [here](https://code.djangoproject.com/ticket/34429#comment:6)).
> 
> I did consider "Set unusable password" but this would require the user to understand the implications, namely, that the user cannot log in using a username and password. That's the root of the wording.
> 
> IMHO, "unsetting" a password could be confused with "set no password". Also, even with an unusable password, you can potentially still log in with a different authentication backend.
> 
> Current text: **Allow login with username and password**

My main concern with this sentence is that, given that Django allows the User model to be customized so that the `USERNAME_FIELD` could be something other than `username`, like email or something system specific, saying `username` here would not be a accurate for those cases.

Also, considering that this flow occurs within the context of the admin, I would suggest to be more direct with the admin user and explicitly saying that the password will be set and it will be unususable. Otherwise I can see admin user scratching their head wondering what's the difference with marking a user as inactive.

> Help text: **If unchecked, the user will have an unusable password and not be able to log in using their username and any password.** (I just realized this had a spelling error)

Same concern about saying `username`,  we could say "credentials" as I suggested?
 
> Warning: **Upon submission, an unusable password will be set which prevents the user from logging in using their username and password. The user's current password will be lost. Potential other means of authentication remain unaffected.**

I would omit "upon submission", I think this is implied from every help text. With that in mind, I would suggest: 

An unusable password will be set which prevents the user from logging in until a new password is set. The user's current password will be lost, though other means of authentication (if configured) remain unaffected.

> I am not hung up on this, but also would like it to be as clear as possible. What's your call?

I proposed a few alternatives above. The summary would be:
* not saying "username and password" explicitly, and
* being clear about how this action is different from making a user inactive

nessita commented:
There is also a potentially awkward user flow when:
1. A user `tesuser1` has an unusable password
2. Later on, the admin user goes to this user's change form to set a valid password:
![image](https://github.com/django/django/assets/124304/ba4dba7e-7f28-4701-a28f-db5e5fcde5b8)
4. The admin user is presented with a "change password" form where no password fields are available:
![image](https://github.com/django/django/assets/124304/8135a66e-ddae-4de6-9263-4c4227b52a2c)

I wonder if we should enable the password fields when going from "change the password using <this form>"?

fsbraun commented:
@nessita Indeed, I initially had the password fields enabled by default. @felixxm requested it differently here: https://github.com/django/django/pull/16942#discussion_r1261060471 

I can revert the change. A third option might be to change the help text if the user does not have a usable password to something like: "Enable this to set a password for this user". 

An update for wording is coming up.

nessita commented:
> @nessita Indeed, I initially had the password fields enabled by default. @felixxm requested it differently here: [#16942 (comment)](https://github.com/django/django/pull/16942#discussion_r1261060471)

Oh, I see. I can understand the goal of having the form properly initialized, but I guess is weird when coming from "no password set". I guess we can proceed as is and then evaluate potential/future ticket reports.

> I can revert the change.

No need for now.

> A third option might be to change the help text if the user does not have a usable password to something like: "Currently this user does not have a usable password. Enable this to set a password for this user".

I like this proposal. May I suggest:

The current password is not usable, check this option to set a new password for user authentication.

fsbraun commented:
@nessita **Thanks once more for helping to clarify all this!** Here's my update. Messages are:

Label:
**Set usable password**

Help text (1/2) for "add user" and "change password" if the current password is usable: 
**An unusable password will render the login credentials unusable.  The user will no longer be able authenticate using a password until a new password is set.** 

Help text (2/2) for "change password" if the current password is unusable: 
**The current password is not usable, check this option to set a new password for user authentication.**

Warning: 
**An unusable password prevents the user from authenticating using a password.
          A user's current password will be lost. Potential other means of authentication
          remain unaffected.**

fsbraun commented on tests/auth_tests/test_forms.py:
```diff
@@ -372,6 +375,15 @@ def test_case_insensitive_username(self):
         )
 
+    def test_unusable_password(self):
+        """Create user with unusable password"""
```
Removed.

fsbraun commented on tests/auth_tests/test_forms.py:
```diff
@@ -372,6 +375,15 @@ def test_case_insensitive_username(self):
+            "usable_password": False,
+        }
+        form = UserCreationForm(data)
+        self.assertTrue(form.is_valid())
```
Test now saves user and double checks username and that the password is not usable.

nessita commented on django/contrib/admin/templates/admin/auth/user/add_form.html:
```diff
@@ -8,3 +8,11 @@
     <p>{% translate "Enter a username and password." %}</p>
   {% endif %}
 {% endblock %}
+{% block extrahead %}
+     {{ block.super }}
+     <script src="{% static 'admin/js/unusable_password_field.js' %}" async></script>
+{% endblock %}
```
```suggestion
{% block extrahead %}
  {{ block.super }}
  <script src="{% static 'admin/js/unusable_password_field.js' %}" async></script>
{% endblock %}
```

nessita commented on django/contrib/auth/forms.py:
```diff
@@ -138,7 +162,11 @@ def _post_clean(self):
         user = super().save(commit=False)
-        user.set_password(self.cleaned_data["password1"])
+        if self.cleaned_data["usable_password"]:
+            password = self.cleaned_data["password1"]
+            user.set_password(password)
```
```suggestion
            user.set_password(self.cleaned_data["password1"])
```

nessita commented on docs/topics/auth/default.txt:
```diff
@@ -1831,7 +1838,12 @@ Creating users
 section of the main admin index page. The "Add user" admin page is different
 than standard admin pages in that it requires you to choose a username and
-password before allowing you to edit the rest of the user's fields.
+password (or setting the password as unusable) before allowing you to edit the rest of
```
I forgot to mention that docs should always be wrapped at 79cols, so this line may need an adjustment.

nessita reviewed (commented):
I did another pass at the code. I would like to find a way to avoid the logic duplication between the base user creation form and the admin password change form. I know the duplication was there already, but with the adding of "usable_password", it feels like we should revisit the code and consider a refactoring.

Perhaps we could consider a "set password mixin" to be used in both form? I can give that a try but is unlikely I can do it before next week.

What do you think?

fsbraun commented:
I'll give it a try. There's an issue with postgres selenium tests which I cannot reproduce.

sarahboyce commented:
> I'll give it a try. There's an issue with postgres selenium tests which I cannot reproduce.

It appears to be an issue with flakiness rather than to do with your change set 👍

nessita commented:
> I'll give it a try. There's an issue with postgres selenium tests which I cannot reproduce.

Thanks! re: selenium, see Sarah's comment (test is flaky).

I'll do another UI review now. If possible, would you also, please, rebase onto main?

nessita commented:
Hey Fabian, I did a UI review as promised. This is looking good! Thank you for being so responsive and open to our feedback.

Before going any deeper in a potential "password change mixin", I'd like to explore a different path for the user details/edit page. I was thinking that we could consider leaving the change password form as is, and instead have a new, simpler, form for just setting the password as unusable.

So, when viewing user details, the admin user is presented with a legend similar to this:

If the user has a usable password: "Raw passwords are not stored, so there is no way to see this user’s password, but you can change the password using <this form>, or disable authentication via <this form>".

If the user has a unusable password: "Re-enable authentication for this user by setting a password using <this form>"

The user creation form will remain as proposed in this PR.

I know we have been going back and forth with this PR, and I'm deeply thankful for your patience, hard work and willingness to adapt. At the same time, I'm evaluating these user experience adjustments as these pages are very high profile and it's important we get this right for both new users and advanced users.

What do you think?

nessita commented:
Hey @fsbraun, I see that your pushed the changes for the password mixin, thanks! Perhaps let's hold on my last UI change proposal until I review this latest revision, I would not want you to keep investing in this without more concrete feedback from me.

Are you in the Discord Django channel? I was wondering if a sync chat would help the (my?) thinking process.

fsbraun commented:
@nessita Just pushed the commit including the Mixin before I saw your thoughts on different forms. Here's my 5 cents on the option to offer two different password change forms:

* On the plus side: I find the text in the user change form explanation texts better. They are also closer to the "competing" active switch for the user. I'd slightly change them to:
  * Raw passwords are not stored, so there is no way to see this user’s password, but you can change the password using <this form>, or disable **password-based** authentication via <this form>"
  * Re-enable **password-based** authentication for this user by setting a password using <this form>
* A potential minus: The click-flow will be different for creating a user and changing a user's password

To me, this is not a clear call. Given the repeated discussion about active users and users with an unusable password, I'd probably nudge to separating the forms.

fsbraun commented:
@nessita Can you share a link for the discord channel?

nessita commented:
> @nessita Can you share a link for the discord channel?

I think so, I'm new to discord so I think I got the right link :-D:

https://discord.gg/WHW9YkFg

I'll be relocating for about an hour, but I'll be online afterwards. I would suggest joining the `#contributor-discussion` channel, where we specifically discuss Django contributions. I'm nessita there as well!

fsbraun commented:
@nessita Thinking a bit more, I come to the conclusion that I would keep the single password change form **and** change the message on the user change form. Something like this:

  * Raw passwords are not stored, so there is no way to see this user’s password, but you can change the password or disable password-based authentication via `<this form>`"
  * Re-enable password-based authentication for this user by setting a password using `<this form>`

fsbraun commented:
Hi @nessita ! I will give it a stab next week.

thibaudcolas commented:
@nessita @fsbraun based on your request, we’ve reviewed this together with Wagtail UI designers & UI developers @benenright, @laymonage, @nicklee and myself. 

Our specific goal was to provide:

> Review IA of the fields + help text as-is in Fabian’s PR
>  Some thoughts on "one form vs. two forms"

---

So their guidance is:

1. Use the same "one form with conditional fields" pattern for both "add" and "edit" for consistency and clarity. With separate "set password" and "remove password" forms, the individual forms would be clearer, but having to choose between forms (or understand there are two forms) would add confusion.
2. On the forms, switch from a "Set password" checkbox to two radio buttons, so both alternatives can be more clearly labelled.
3. As far as field order, makes sense for the "Set password" field to be where it is now.
4. The "set password: true" should always be selected by default, as that’s the most likely intent for both a new user, and editing an existing user regardless of that user’s current status.

And additionally, change the wording and placement of this "this form" link so it’s more obvious what the link takes you to. Possibly use a button rather than a link in text, with that button label being clear about the form’s purpose.

I didn’t ask them too much input on exact wording but they did say:

- The help text needs work
- Consider terms like "Password-less" or "Other authentication mechanism"
- Even better if we could add "most common use cases" scenarios as examples in the help text, like "Use this for SSO" or similar

I’ll try to get back to the wording on my own based on their input and propose specific tweaks.

---

Congratulations @benenright & @nicklee on a first contribution to Django 😉

fsbraun commented:
Thank you so much, @thibaudcolas, @benenright, and @nicklee!

smithdc1 commented on django/contrib/auth/forms.py:
```diff
@@ -194,6 +238,11 @@ def __init__(self, *args, **kwargs):
+            if self.instance and not self.instance.has_usable_password():
+                password.help_text = _(
+                    "Enable password-based authentication for this user by setting "
+                    'a password using <a href="{}">this form</a>.'
```
https://design-system.service.gov.uk/components/text-input/#hint-text

What do you think about avoiding links in help text?

> Do not include links within hint text. While screen readers will read out the link text when describing the field, they will not tell users that the text is a link.

fsbraun replied:
@smithdc1 I personally think this should be a **button** but did not want to interfere with #17419.

nessita replied:
@smithdc1 I kind of agree with you but if we remove the link from the help text, how/where can we expose the change password functionality? Any ideas or suggestions? (in any case, I would do that in a separated ticket/PR)

fsbraun replied:
@nessita I'd make it a button, just blow the help text. It'll be a simple string change that can be easily added to this PR. It would look like this: 
<img width="694" alt="image" src="https://github.com/django/django/assets/16904477/eeee524b-3845-411a-a804-7217d8778e62">
I can add this here or create a separate mini-PR. What's your preference?

thibaudcolas replied:
I think this would be an amazing change. Whether it’s a separate PR or not – not my call.

thibaudcolas replied:
I think this would be an amazing change. Whether as a separate PR or not – not my call.

fsbraun replied:
@smithdc1 @thibaudcolas I created this PR #17489 to introduce the set/change password button.

nessita commented on docs/releases/5.1.txt:
```diff
@@ -32,7 +32,9 @@ Minor features
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-* ...
+* When adding a new user or changing a password in the admin UI, the
+  administrator can chose to set an unusable password effectively disabling
+  a user from logging in with a password.
```
```suggestion
* Enhanced User's password management now allows administrators to enable or
   disable password-based authentication for any user by setting a usable or
   unusable password. This applies to both to user creation and user change
   forms.
```

nessita replied:
I think this would be best placed under the `django.contrib.auth` subsection. Also, I'd suggest a rewording as per:
```suggestion
* Improved password management in the User model now allows admins to enable or
  disable password-based authentication by setting a usable or unusable        
  password, applicable in user creation and change forms.                      
```

nessita commented on docs/topics/auth/default.txt:
```diff
@@ -1632,7 +1632,14 @@ provides several built-in forms located in :mod:`django.contrib.auth.forms`:
 .. class:: AdminPasswordChangeForm
 
-    A form used in the admin interface to change a user's password.
+    A form used in the admin interface to change a user's password. This form
+    also allows the password to be set as an :meth:`unusable password
+    <django.contrib.auth.models.User.set_unusable_password>` effectively
+    barring the user from logging in.
```
```suggestion
    A form used in the admin interface to change a user's password, including
    the ability to set an :meth:`unusable password
    <django.contrib.auth.models.User.set_unusable_password>`, effectively
    barring the user from logging in using a password-based authentication.
```

nessita commented on docs/topics/auth/default.txt:
```diff
@@ -1845,7 +1852,12 @@ Creating users
+password (or setting the password as unusable) before allowing you to edit the
+rest of the user's fields.
+
+    .. versionchanged:: 5.1
+
+        Checkbox option to set an unusable password was added.
```
```suggestion
    .. versionchanged:: 5.1

        Option to set an unusable password was added.
```

nessita reviewed (commented):
I've done yet another UX/string review. Thank you for the changes! :balloon: 

In all honesty I don't like radio buttons for when there is a True/False choice, but I'll abide with the advice of the design experts. With that in mind, I most certainly would change:

* "_Password-based login_" to be consistently "_Password-based authentication_" (and ensure that we use this wording everywhere, when referencing the same thing)
* The radio buttons labels for this option should be "_Enabled_" and "_Disabled_", so:
```
Password-based Authentication: [X] Enabled [ ] Disabled
```
* The sentence "_Potential other authentication mechanisms such as Single Sign-On remain unaffected._ " should be rephrased to:
"_Other authentication mechanisms, such as Single Sign-On or LDAP, remain unaffected._"
* I would suggest that we remove both help texts for the new radio button field, because I think they pollute the UI without adding much clarity (see next point).
* I would slightly change the phrase "_Not setting a password will render the login credentials unusable. The user will no longer be able authenticate using a password until a new password is set._"  to:
"_Not setting a password will render the login credentials unusable. The user will not be able to authenticate using a password until a new password is set._" and use this as the yellow warning main text when "Password-base Authentication" is **disabled**.
* I would join all warning sentences into one paragraph except for "_The user's current password will be lost._", which I think needs to be a separated, red-like-warning message, because is a destructive action.

@fsbraun if you could work on the changes above, that would be great, otherwise no worries, I can do those myself next week.

thibaudcolas commented on django/contrib/auth/forms.py:
```diff
@@ -180,7 +219,7 @@ class UserChangeForm(forms.ModelForm):
         help_text=_(
             "Raw passwords are not stored, so there is no way to see this "
-            "user’s password, but you can change the password using "
+            "user’s password, but you can change or unset the password via "
```
👍 seems like a reasonable adjustment to me.

thibaudcolas commented on django/contrib/admin/templates/admin/auth/user/unusable_password_warning.html:
```diff
@@ -0,0 +1,19 @@
+  <ul id="id_unusable_warning" class="messagelist">
+    <li class="warning">
+      {% blocktranslate %}
+        Not setting a password will render the login credentials unusable.
+        The user will not be able to authenticate using a password until a
+        new password is set. Other authentication mechanisms, such as
+        Single Sign-On or LDAP, remain unaffected.
```
I think the first sentence just adds confusion so should definitely be removed. Starting with "not" makes it tough to read, and "login credentials" will be lost on people. So at least, do this:

```suggestion
        The user will not be able to authenticate using a password until a
        new password is set. Other authentication mechanisms, such as
        Single Sign-On or LDAP, remain unaffected.
```

I’d go further and use even simpler language. Don’t care too much about the exact wording but the specific changes:

- "authenticate" / "authentication mechanism" -> "log in"
- "remain unaffected" -> needs to use simpler language too.

Or even simpler:

```suggestion
        The user will not be able to log in with a password.
        They may still be able to log in through other options, such as Single Sign-On or LDAP.
```

nessita replied:
@thibaudcolas the thing is that "log in" is not (strictly) [the same](https://english.stackexchange.com/questions/80862/difference-between-log-in-authenticate-and-sign-in) as "authenticate". One could authenticate via other means (think about [Basic Authentication](https://en.wikipedia.org/wiki/Basic_access_authentication)) and still not do a "log in" flow.

This new UI affects how authentication works, not necessarily the login process.

thibaudcolas reviewed (changes requested):
Just reviewing the copy and nothing else – seems like a clear improvement to me over previous iterations. I’d suggest going further. I’d really like to see changes to `unusable_password_warning.html`. For everything else here – I’d suggest going further but I don’t feel strongly about it.

## Field label

Currently this is:

- Password-based authentication
    - 🔘 Enabled
    - ⭘ Disabled

I think this is an improvement over the previous "Set password" checkbox, but we could go further and make more use of having descriptive labels.

- Authentication type
    - 🔘 Password
    - ⭘ Other (no password set)

## Unusable password

I don’t feel strongly about this – but we did discuss that this was just an implementation detail, so I question whether it should even be in the release notes / docs, aside from the method name.

## "Change password" form title

Not a big deal for me but I don’t get why we would want different form titles for those two options, as the form technically allows both types of changes. Why not always a generic "Set password", or an open-ended "Change or set password"?

One exception would be the form’s submit button, where I think it’s great we’re swapping depending on what the form would do in its current state.

fsbraun commented:
@thibaudcolas Thanks so much! I pushed an improved `unusable_password_warning.html` based on your second suggestion (keeping the term "authenticate" instead of "log in" as @nessita earlier pointed out the need for consistent language). It now reads:

```
The user will not be able to authenticate using a password. 
They may still be able to authenticate using other options, 
such as Single Sign-On or LDAP.
```

Also, I relabelled the radio buttons to
```
Authentication type:   [X] Password   [ ] Only other (no password)
```
I was struggling a bit to avoid the impression that you'd have to select "Other" for other authentication backends to work, while in fact they are independent.

nessita commented:
> ```
> Authentication type:   [X] Password   [ ] Only other (no password)
> ```
> I was struggling a bit to avoid the impression that you'd have to select "Other" for other authentication backends to work, while in fact they are independent.

@thibaudcolas @fsbraun Precisely for this reason,  I find `Other` confusing. OTOH, `Only other`, while more accurate, feels cryptic and that it conveys an effect with unclear side-effects.

In my view, we should use as context the default authentication backend, avoiding mixed references with other backends that may or may not exist. This prevents confusion, especially when users may not be aware of other backends. Even more, an admin user should be able to disable password-based authentication altogether, even if no other auth backend is configured, so saying `Other`  could hint that you can't do that *unless* you define other auth backend.

I stand by my original suggestion which limits the messaging to the authentication backend that makes use of the password:

```
Password-based authentication:  [x] Enabled  [ ] Disabled
```

thibaudcolas commented:
👍 fair enough :)

fsbraun commented:
@thibaudcolas Great! I've reverted the radio buttons as @nessita suggested. 🚀. From my side, this is read for take-off 😎.

nessita commented:
> @thibaudcolas Great! I've reverted the radio buttons as @nessita suggested. 🚀. From my side, this is read for take-off 😎.

Understood, will review again soon! Thanks for your constant work and amazing predisposition.

fsbraun commented on django/contrib/auth/forms.py:
```diff
@@ -89,28 +89,86 @@ def widget_attrs(self, widget):
+            user.save()
+            if hasattr(self, "save_m2m"):
+                self.save_m2m()
+        return user
```
Great idea!

fsbraun commented on django/contrib/admin/templates/admin/auth/user/change_password.html:
```diff
@@ -11,10 +11,14 @@
 &rsaquo; <a href="{% url opts|admin_urlname:'changelist' %}">{{ opts.verbose_name_plural|capfirst }}</a>
 &rsaquo; <a href="{% url opts|admin_urlname:'change' original.pk|admin_urlquote %}">{{ original|truncatewords:"18" }}</a>
-&rsaquo; {% translate 'Change password' %}
+&rsaquo; {% if form.has_usable_password %}{% translate 'Change password' %}{% else %}{% translate 'Set password' %}{% endif %}
```
Added a regression test.

fsbraun commented:
@nessita I made the changes, but kept the JS as discussed.

fsbraun commented:
@nessita Any news on this?

nessita commented on django/contrib/admin/templates/admin/auth/user/unusable_password_warning.html:
```diff
@@ -0,0 +1,18 @@
+        {% translate "The user's current password will be lost. " %}
+      </li>
+    {% endif %}
+
```
Could we remove this empty line?
```suggestion
```

fsbraun replied:
Whole file is gone now... 😉. It is not needed any more if CSS manages what is viewable.

nessita commented on django/contrib/admin/templates/admin/auth/user/unusable_password_warning.html:
```diff
@@ -0,0 +1,18 @@
+    <li class="warning">
+      {% blocktranslate %}
+        The user will not be able to authenticate using a password.
+        They may still be able to authenticate using other options,
```
```suggestion
        They may still be able to authenticate using other backends,
```

nessita commented on django/contrib/auth/forms.py:
```diff
@@ -180,7 +220,7 @@ class UserChangeForm(forms.ModelForm):
         help_text=_(
             "Raw passwords are not stored, so there is no way to see this "
-            "user’s password, but you can change the password using "
+            "user’s password, but you can change or unset the password via "
```
```suggestion
            "user’s password, but you can change or unset the password using "
```

fsbraun replied:
Changed.

nessita commented on docs/releases/5.1.txt:
```diff
@@ -44,6 +43,10 @@ Minor features
 
 * The default iteration count for the PBKDF2 password hasher is increased from
   720,000 to 870,000.
+* Improved password management in the User model now allows admins to enable or
```
```suggestion

* Improved password management in the User model now allows admins to enable or
```

fsbraun replied:
Done.

nessita commented on docs/releases/5.1.txt:
```diff
@@ -32,7 +32,6 @@ Minor features
 :mod:`django.contrib.admin`
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-* ...
```
I think we need to restore this 3 dots.

nessita commented on docs/releases/5.1.txt:
```diff
@@ -44,6 +43,10 @@ Minor features
+* Improved password management in the User model now allows admins to enable or
+  disable password-based authentication by setting a usable or unusable
+  password, applicable in user creation and change forms.
+
```
Only one empty line afterwards
```suggestion
```

nessita commented on docs/topics/auth/default.txt:
```diff
@@ -1632,7 +1632,14 @@ provides several built-in forms located in :mod:`django.contrib.auth.forms`:
+
+    .. versionchanged:: 5.1
+
+        Option to disable password-based authentication was added.
```
```suggestion
        Option to disable (or re-enable) password-based authentication was added.
```

nessita commented on docs/topics/auth/default.txt:
```diff
@@ -1845,7 +1852,12 @@ Creating users
+
+    .. versionchanged:: 5.1
+
+        Option to disable password-based authentication was added.
```
```suggestion
        Option to disable (or re-enable) password-based authentication was added.
```

nessita commented on tests/admin_views/test_password_form.py:
```diff
@@ -0,0 +1,73 @@
+
+        # Warning disappears
+        with self.assertRaises(StaleElementReferenceException):
+            self.assertFalse(warning.is_displayed())
```
This test is great, we'd need a couple more for:
* existing user starting with usable password
* existing user starting with non usable password

fsbraun replied:
Done.

nessita commented on tests/admin_views/test_password_form.py:
```diff
@@ -0,0 +1,73 @@
+        )
+
+        # For add user the default is a usable password
+        self.assertTrue(pw_switch_on.is_selected())
```
From [the contributing guide](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/), in general is preferred to use:
```suggestion
        self.assertIs(pw_switch_on.is_selected(), True)
```

nessita commented on tests/admin_views/test_password_form.py:
```diff
@@ -0,0 +1,73 @@
+        self.assertFalse(password1.is_displayed())
+        self.assertFalse(password2.is_displayed())
+
+        # Finally check that the warning is displayed
```
I think we should also be testing for the expected button legend, since we are changing it via javascript?

fsbraun replied:
Since JS is gone now, this can be checked by the Python tests.

fsbraun replied:
… and just pushed.

nessita commented on tests/auth_tests/test_views.py:
```diff
@@ -1440,14 +1440,35 @@ def test_user_change_password(self):
+        self.assertNotContains(response, "<h1>Set password")
+        self.assertNotContains(response, "&rsaquo; Set password")
+
+        # Test for title with unusable password with a test user
```
Could we split this in two tests? one starting with a user with a usable password, and the other for a user without.

fsbraun replied:
Done.

fsbraun commented on django/contrib/admin/templates/admin/auth/user/unusable_password_warning.html:
```diff
@@ -0,0 +1,18 @@
+    </li>
+    {% if form.user.has_usable_password %}
+      <li class="error">
+        {% translate "The user's current password will be lost. " %}
```
File is gone. Changed here: 
https://github.com/fsbraun/django/blob/7c1e0b7c16b079b6f3311eed94d48784af697d5e/django/contrib/auth/forms.py#L101-L106

ngnpope commented on tests/admin_views/tests.py:
```diff
@@ -7192,6 +7192,7 @@ def test_save_button(self):
             reverse("admin:auth_user_add"),
             {
                 "username": "newuser",
+                "usable_password": True,
```
Following on from my comment earlier, this is relying on `str(True)` rather than an explicit value. As these are values of a query string, I'd expect them to be in serialized form, i.e. `"True"`, that matches the actual value used in the form choices as defined in `SetPasswordMixin`. _(Although maybe `"1"` as for `_continue` below would be more conventional/consistent?)_

fsbraun replied:
I'd go for `"true"`; otherwise we have `[value="0"]` and `[value="1"]` as CSS selectors which are not very explicit.

ngnpope reviewed (changes requested):
Thanks for your continued efforts on this one. It's been a while since I dropped in - this is shaping up to be a very nice improvement 👍🏻

ngnpope commented on django/contrib/admin/static/admin/js/unusable_password_field.js:
```diff
@@ -0,0 +1,25 @@
+"use strict";
+// Fallback JS for browsers which do not support :has selector used in
+// admin/css/unusable_password_fields.css
+// Remove file once all supported browsers support :has selector
+try {
+    // If browser does not support :has selector this will raise an error
+    document.querySelector("form:has(input)");
+} catch (error) {
```
But I'll defer to @nessita's decision on this 😄

felixxm replied:
I disagree, what you're proposing is a really aggressive policy :fire: . We cannot drop support for 1 months old Firefox releases just because the latest one supports `has()`. I (for example) have 119 locally :older_adult:. We need to remember that admin is for app users and not all of them bump Firefox daily.

ngnpope replied:
I was looking at this with a lens of what the state of things will look in 8 months time rather than right now. In that regard we'll [be expecting](https://whattrainisitnow.com/calendar/) v129 to be the latest version[^1]. _(Also v128 will be the next ESR and, yes, there'll be one more release of v115 ESR to come in September before it hits end-of-life in October.)_

Our documentation states "recent modern versions" and I get that this is quite nebulous, but I would hardly call what I proposed "really aggressive" in light of that. I will concede though that, in this case, it's likely to cause more than "minor stylistic differences" to the feature if the browser doesn't have the required support.

Regardless, I was just making a case that including this seemed pointless and, if that's rejected, it's fine ❤️

[^1]: I certainly hope you've upgraded to something more recent than v119 by then for your own security 😛

nessita replied:
I have thoroughly tested the three paths:
1. Browser with `has` support
2. No `has` support and JS enabled
3. No `has` support and JS disabled

User experience is quite decent in all cases (after I made some tweaks to the forms and the view flows). I think we can safely land as is and cleanup in a future iteration. Thank you!

fsbraun commented on django/contrib/auth/forms.py:
```diff
@@ -89,50 +89,112 @@ def widget_attrs(self, widget):
-            ] = True
+        warnings = f'<li class="warning">{self.unusable_warning}</li>'
+        if extra_warning:
+            warnings += f'<li class="error">{extra_warning}</li>'
```
The semantically correct levels would be "info" and "warning". But info has a checkbox as an icon, which I find misleading. But that's representation. It's probably best to stay with the semantics. @nessita What do you think?

fsbraun commented on django/contrib/auth/forms.py:
```diff
@@ -89,50 +89,112 @@ def widget_attrs(self, widget):
+        label=_("Password-based authentication"),
+        required=False,
+        initial="True",
+        choices=(
+            ("True", _("Enabled")),
+            ("False", _("Disabled")),
+        ),
```
`"true"`/ `"false"` make sense to me and I would prefer it to 1 and 0.

nessita commented:
Hello @fsbraun! I'm back to this review :muscle: Details below for UI review, code review coming next.

As discussed in Discord, I agree that the need for JS is (at this point) justified and we'll have to wait to 5.2 to remove this JS and then rely solely on the `:has()` CSS pseudo-class as you suggested.

I can make the changes requested below if you are busy, let me know! Thanks again :trophy: 

UI review
-------------
* The separation between the radio buttons feels weird, the options are too apart from eachother. Is there something that can be done?  
![image](https://github.com/django/django/assets/124304/e4a5741a-ad53-4bf6-97eb-6f2a4869bfc1)

* In the message shown in the screenshot, I see two issues with how it is displayed right now:
  1. the left margin/padding makes it looks a bit unpolished, and
  2. the fact that is green (usually meaning success) is a bit unexpected for the tone of the message. Could we instead make it a "traditional" help text with its usual formatting?
![image](https://github.com/django/django/assets/124304/cd019d19-ac40-4b7d-8ace-c028ca96622b)

What about something like:
![image](https://github.com/django/django/assets/124304/75e6fe8e-c8bb-4471-871e-ee2ff672a1cf)

With diff:
```diff
diff --git a/django/contrib/auth/forms.py b/django/contrib/auth/forms.py
index a5d7f8017b..f2978bf992 100644
--- a/django/contrib/auth/forms.py
+++ b/django/contrib/auth/forms.py
@@ -98,9 +98,9 @@ class SetPasswordMixin(forms.Form):
     error_messages = {
         "password_mismatch": _("The two password fields didn’t match."),
     }
-    unusable_warning = _(
-        "The user will not be able to authenticate using a password. "
-        "They may still be able to authenticate using other backends, "
+    usable_help_text = _(
+        "Whether the user will be able to authenticate using a password or not. "
+        "If disabled, they may still be able to authenticate using other backends, "
         "such as Single Sign-On or LDAP."
     )
 
@@ -112,6 +112,7 @@ class SetPasswordMixin(forms.Form):
         widget=forms.RadioSelect(
             attrs={"class": "radiolist inline"},
         ),
+        help_text=usable_help_text,
     )
     password1 = forms.CharField(
         label=_("Password"),
@@ -128,17 +129,14 @@ class SetPasswordMixin(forms.Form):
         help_text=_("Enter the same password as before, for verification."),
     )
 
-    def __init__(self, *args, **kwargs):
-        extra_warning = kwargs.pop("extra_warning", None)
+    def __init__(self, *args, extra_warning=None, **kwargs):
         super().__init__(*args, **kwargs)
-        warnings = f'<li class="info">{self.unusable_warning}</li>'
         if extra_warning:
-            warnings += f'<li class="warning">{extra_warning}</li>'
-        self.fields[
-            "usable_password"
-        ].help_text = (
-            f'<ul id="id_unusable_warning" class="messagelist">{warnings}</ul>'
-        )
+            self.fields[
+                "usable_password"
+            ].help_text += (
+                f'<ul id="id_unusable_warning" class="messagelist">'
+                f'<li class="warning">{extra_warning}</li></ul>')
 
     def clean_password1(self):
         password1 = self.cleaned_data.get("password1")
@@ -528,7 +526,7 @@ class AdminPasswordChangeForm(SetPasswordMixin, forms.Form):
     """
 
     required_css_class = "required"
-    password_lost_warning = _("The current password for this user will be lost.")
+    password_lost_warning = _("If disabled, the current password for this user will be lost.")
 
     def __init__(self, user, *args, **kwargs):
         self.user = user
```

User workflow review
------------------------------
1. When javascript is disabled, on user add form, if password-based authentication is disabled, the values from the password fields should be cleaned in a way where no validation is done but form errors are returned indicating that fields should be empty in this case. Right now, when submitting the form with password auth disabled **and** some values in the password forms, the user get form errors regarding the password fields.
2. Password-based authentication value on forms for existing users should default to the correct/set value (currently, visiting "change password" form with a user with unusable password shows a form with "Enabled" chosen by default).
4. Even with javascript enabled, I see these two buttons for an existing user with non usable password:
![image](https://github.com/django/django/assets/124304/46d6055a-de3d-413e-ba6f-09ba476f916d)

fsbraun commented:
@nessita Thank you very much! A few remarks/questions:
* The spacing between radio buttons is defined for all forms by the admin style set. I would not change it for this specific form. Maybe that's another thing to change in a separate PR for all of django admin's radio buttons?
* I'm fine with the "success" message going into the help_text. ☑️ **Can you commit your patch?**
* Good catch that the hidden password fields can cause a **form validation failure. I will update the PR accordingly**.
* If a user has an unset password, clicking "set password" should IMHO not prompt them to unset it again. We discussed this here: https://github.com/django/django/pull/16942#issuecomment-1717755980 and I already changed it back and forth. I suggest not changing this.
* How did you create the scenario #3 in user workflow (the two submit buttons visible)? I tried and cannot recreate it even with Firefox 118.

I will check the form validation.

nessita commented:
> @nessita Thank you very much! A few remarks/questions:
> 
>     * The spacing between radio buttons is defined for all forms by the admin style set. I would not change it for this specific form. Maybe that's another thing to change in a separate PR for all of django admin's radio buttons?

That's fair, and I agree, let's leave this as is, using the shared CSS style.

>     * I'm fine with the "success" message going into the help_text. ☑️ **Can you commit your patch?**

Yes, will do that next, along with a `main` rebase.
 
>     * Good catch that the hidden password fields can cause a **form validation failure. I will update the PR accordingly**.

Yey!

>     * If a user has an unset password, clicking "set password" should IMHO not prompt them to unset it again. We discussed this here: [Fixed #34429 -- Allowed setting unusable password via admin UI. #16942 (comment)](https://github.com/django/django/pull/16942#issuecomment-1717755980) and I already changed it back and forth. I suggest not changing this.

You are absolutely right, sorry for the back and forth. I guess we could (eventually) ask for a second opinion to Thibaud, but we may do this when we are closer to landing.
 
>     * How did you create the scenario no. 3 in user workflow (the two submit buttons visible)? I tried and cannot recreate it even with Firefox 118.

All I did was:
1. Add a user with unusable password, submit
2. Visit this user's details page
3. Click on the "set the password using this link"
4. Two buttons are shown (this is NOT the case for a user with a previously usable password)

Does this help?

> I will check the form validation.

Excellent! Thank you.

fsbraun commented:
@nessita I just pushed an update. Now, the test for password1 and password2 matching is skipped if an unusable password is to be set.

I cannot, however, reproduce the situation with the two submit buttons. Their visibility should be managed by the JS polyfill. Which browser are you seeing this? Is there any message on the JS console? Do both buttons stay if you flip the radio buttons for a usable password? Does this persist if you empty the browser's cache?

nessita commented:
Hello @fsbraun! I pushed my changes though the selenium tests need fixing one we decide on the final flow.

Regarding the two buttons for changing the password for a user with unusable password, this is working on Chrome Version 114.0.5735.198 but is not working on Firefox 120.0.1. I tested this on a private window, there errors in the developer console as shown below:
```
Uncaught TypeError: document.getElementById(...) is null
    <anonymous> http://localhost:9000/static/admin/js/unusable_password_field.js:17
    <anonymous> http://localhost:9000/static/admin/js/unusable_password_field.js:23
    <anonymous> http://localhost:9000/static/admin/js/unusable_password_field.js:10
```
See screenshot:
![image](https://github.com/django/django/assets/124304/b050c190-71c6-4d04-bf0b-a9b2b383b4b8)

nessita commented:
Double buttons on Firefox is still present even if browser cache is clear.

fsbraun commented:
@nessita Ah, I see now: The double buttons are a result of your replacing the warnings with the help text: 
* When the polyfill tries to find the warning to set its visibility, it fails. 
* Probably the same for the selenium tests: They must not test that the help text is hidden if it is supposed to be visible all the time.

I can quickly fix both, but am unsure what decision has to be made on the "final flow". Will push the fix and wait for further feedback or instructions.

nessita commented:
Hey @fsbraun, I'm doing another review now.

In order to help with the review (for myself, and for others), I started isolating the creation of the `SetPasswordMixin` into a logical unit (separated commit). I've found a potential blocker, this line I'm highlighting is removed from `AdminPasswordChangeForm` and not added anywhere:

```
password_validation.validate_password(password2, self.user)
```

Also,  I've noticed that a rebase onto main causes conflicts because older commits cause conflicts which are later fixed by newer commits. Given that there are 71 commits, would you agree to squash commits into isolated logical units? This way, it's easier to review newer commits, and also we'll comply with [making the commit history as usable as possible](https://docs.djangoproject.com/en/5.0/internals/contributing/committing-code/#handling-pull-requests).

Following these two items, would you have time/interest in creating a separated commit (with commit message suffix `Refs #34429`) that refactors the common logic for setting a password into `SetPasswordMixin`?
We would need to evaluate how to add this lost line back in and add a few tests for it (since no test fails when removed). The tests adding would ideally be done in its own isolated PR, to be landed as soon as ready and separated from this work.

Let me know! Thank you.

fsbraun commented:
@nessita I added back the missing line and regression tests for the future for both creating users and changing a user's password. This should remove the blocker. The set of valid form inputs stays unchanged, but the validation errors shown to the user now are consistent across all the forms again. The relevant edge case is an empty password1 and a non-empty non-validating password2.

I'm strongly supporting to squash all commits of this PR upon merging. With the amount of back and forth we have had in this PR, I doubt looking at single commits will be in any way illuminating. 

I just merged the changes from main back into the PR. So there should not be any conflicts (though I cannot see this in GitHub).

Do we really have to split the PR into two commits? I'd be fine with a single squashed commit. Certainly for me, the procedure would cost considerable time and probably also be error-prone. The benefits to me seem not so great in comparison, as the result is a bit artificial: The mixin solely is a tool to not repeat oneself.

nessita commented:
> @nessita I added back the missing line and regression tests for the future for both creating users and changing a user's password. This should remove the blocker. The set of valid form inputs stays unchanged, but the validation errors shown to the user now are consistent across all the forms again. The relevant edge case is an empty password1 and a non-empty non-validating password2.

Great, thank you!

> I'm strongly supporting to squash all commits of this PR upon merging. With the amount of back and forth we have had in this PR, I doubt looking at single commits will be in any way illuminating.

I beg to differ, though I understand your view. I can imagine you may find the ongoing round of reviews and sometimes mildly contradictory requests a bit repetitive.
I want to say again that we appreciate your patience, and that PR review is an iterative process. Django source code is of extremely high quality both in terms of correctness and git history, and as reviewers, we need to ensure that it continues that way.

> I just merged the changes from main back into the PR. So there should not be any conflicts (though I cannot see this in GitHub).

Thanks! For future reference, it'd be preferred to do a rebase instead of a merge (`git rebase main`), that way the PR commits are replayed on top of latest `main` and is a bit cleaner/easier to review and later to merge.

> Do we really have to split the PR into two commits?

Yes, I think so.

> I'd be fine with a single squashed commit. Certainly for me, the procedure would cost considerable time and probably also be error-prone. The benefits to me seem not so great in comparison, as the result is a bit artificial: The mixin solely is a tool to not repeat oneself.

I understand if you'd prefer not to tackle this yourself. However, to maintain commits as isolated logical units and anticipate a potential future need for a revert, I believe the mixin refactor (without any logic change) should serve as a sort of prerequisite change. As mentioned in my previous comment, (eventually) consolidating all the changes into a single commit can make the review process more convoluted, cumbersome, and error-prone—it's easier to overlook something.

Because of the above, my plan is to produce and push these 3 commits as time permits:

1. isolate the newly added (missing in main) tests for the password management form
2. do the mixin
3. the actual ticket changes 

Reorganizing these commits will significantly facilitate the review process. It's more straightforward (at least in my experience) to complete reviews for each logical unit, potentially landing those, and then concentrate on the remaining changes with a more focused approach (the classic divide and conquer).

nessita commented:
> @nessita OK! I just feel not sure how to do that exactly. It might be my limited mastery of git that scares me. (I just clicked update branch on my fork in GitHub, for example. Next time I will run `git rebase main`.)

Not being sure is completely fine! Also do note that you need to update your `main` before doing the rebase.

The big advantage for using rebasing (IMHO) is that your commits always end up at the "top of the stack of commits". For example, `git rebase <ref>` means (I'll handwave a bit, there are tons of articles out there explaining this with more technically correct terms):

* all your commits since `<ref>` are put aside
* repo is updated with latest changes for `<ref>` (this is why you need to update `main` before rebasing)
* the commits are re-played on top of the latest changes 

Doing this consistently and often, ensures that your branch history looks like this:
```
main-hash-1
main-hash-2
...
main-hash-N
your-branch-hash-1
your-branch-hash-2
...
your-branch-hash-M
```
When newer revnos are landed in main, say revnos `main-hash-N+1` and `main-hash-N+2`, after a rebase your branch looks like this:
```
main-hash-1
main-hash-2
...
main-hash-N
main-hash-N+1
main-hash-N+2
your-branch-hash-1
your-branch-hash-2
...
your-branch-hash-M
```

To the contrary, when doing merges from main into your branch, your history starts looking like this:
```
main-hash-1
main-hash-2
...
main-hash-N
your-branch-hash-1
your-branch-hash-2
...
your-branch-hash-M
your-branch-hash-M+1
```

The difference is that `your-branch-hash-M+1` is a single merge commit of both revnos from main ( `main-hash-N+1` and `main-hash-N+2` merged into a single commit). So your branch history incorporates newer changes from `main` but starts diverging quite considerably. 

I can go on and on about this, but I figured you could continue the investigation on your own with these pointers. 

> Isolating the changes to the tests should be easy. There's probably some thinking involved when separating out the mixin.

I agree. I have separated and simplified the test in PR #17729. I removed the testing on both values of password1 because as far as I understood, the fact that password1 would be empty or not was irrelevant for the call:

`password_validation.validate_password(password2, self.user)`

If I'm missing something here, please let me know.

> Out of curiosity, how do you do this? You would need to "undo" all commits while keeping the changes and then redoing the changes in a new order.

What I did was doing multiple *interactive* rebases. An interactive rebase is a rebase as described above but before replaying the commits from your branch, you get a screen to "interact" with git and instruct it to do things, perhaps differently, with the commits that is about to replay.

Specifically for this PR, which had 76 commits before my work, the rebasing was very painful and slow. It took me almost 2 hours to resolve all the conflicts that were arising after every commit from the 76 pending ones was applied.

I think I got the squash and rebase correct. Will push the result in a few.

nessita commented:
One thing that I noticed while rebasing and fixing conflicts is that the adding in various places (tests) of `"usable_password": "true"` to form data could be removed, since that's the default and I don't think is necessary to set it explicitly in tests that are not relevant to this feature. Let me know if there is a reason to add it!

nessita commented:
If you wanna give it a shot to splitting the Mixin into its own commit, [this article](https://dev.to/timmouskhelichvili/how-to-split-a-git-commit-into-multiple-ones-3g6f) may help when it comes to finding the necessary git magic incantations.

felixxm commented on django/contrib/auth/forms.py:
```diff
@@ -89,50 +89,108 @@ def widget_attrs(self, widget):
+        return user
+
+
+class BaseUserCreationForm(SetPasswordMixin, forms.ModelForm):
```
This class is [documented](https://docs.djangoproject.com/en/dev/topics/auth/default/#django.contrib.auth.forms.BaseUserCreationForm), and its documentation no longer matches the implementation.

fsbraun replied:
@felixxm Thanks for pointing this out. I've updated the documentation.

felixxm commented on docs/topics/auth/default.txt:
```diff
@@ -1632,7 +1632,15 @@ provides several built-in forms located in :mod:`django.contrib.auth.forms`:
+
+    .. versionchanged:: 5.1
+
+        Option to disable (or re-enable) password-based authentication was
```
`reenable` is even in the `spelling_wordlist`:
```suggestion
        Option to disable (or reenable) password-based authentication was
```

felixxm commented on tests/admin_views/test_password_form.py:
```diff
@@ -0,0 +1,145 @@
+        )
+
+    def test_add_user_unusable_password_css(self):
+        """Selecting and deselecting the unusable password field shows/hides the
+        fields and the warning when adding a user"""
```
We don't wrap docstrings this way.

felixxm commented on tests/admin_views/test_password_form.py:
```diff
@@ -0,0 +1,145 @@
+        # Switch usable_password selector
+        pw_switch_off.click()
+
+        # Check the click did switch the checkbox
```
To add more explanation, docstrings should state the expected behavior and omit prefixes like "Check" since all tests check things. This guideline is from [Python coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style).

felixxm commented on tests/auth_tests/test_views.py:
```diff
@@ -1440,14 +1440,24 @@ def test_user_change_password(self):
         self.assertEqual(urljoin(user_change_url, rel_link), password_change_url)
 
+        response = self.client.get(password_change_url)
+        # Test the form title with original (usable) password
```
To add more explanation, comments should state the expected behavior and omit prefixes like "Test the" since all tests test things. This guideline is from [Python coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style).

fsbraun commented:
@nessita Did I do this OK?

nessita commented:
> @nessita Is this what you expected?

Yes, quite close! Could you please adjust the Mixin docstring to not mention the unusable password bit in the first commit? That piece should only be added in the second commit.

fsbraun commented:
@nessita There you go!! 😎😃

Thanks to @thibaudcolas for his patient introduction to interactive rebase.

nessita commented:
Hey @fsbraun, thanks for your patience. Last week and today I've been working on final tweaks for this PR. I have reviewed logic and tests in detail, making some adjustments to the UI and UX, and adding more tests to ensure proper coverage of changes.

Tomorrow I'll give the docs another pass and hopefully I will approve and merge. :crossed_fingers:

nessita commented:
I believe this is ready for merge, will do so later today after the CI runs complete.
